http-clients/apache-client/src/main/java/software/amazon/awssdk/http/apache/internal/conn/IdleConnectionReaper.java

```diff
@@ -15,181 +15,163 @@
-     * state. Any use of a service client will cause the thread to be restarted.
-     *
-     * @return true if an actual shutdown has been made; false otherwise.
+     * @return The singleton instance of this class.
      */
-    public static synchronized boolean shutdown() {
-        if (instance != null) {
-            instance.markShuttingDown();
-            instance.interrupt();
-            CONNECTION_MANAGERS.clear();
-            instance = null;
-            return true;
-        }
-        return false;
+    public static IdleConnectionReaper getInstance() {
+        return INSTANCE;
     }
 
-    /**
-     * For testing purposes.
-     * Returns the number of connection managers currently monitored by this
-     * reaper.
-     */
-    static int size() {
-        return CONNECTION_MANAGERS.size();
+    private synchronized void setupExecutorIfNecessary() {
+        if (exec != null) {
+            return;
+        }
+
+        ExecutorService e = executorServiceSupplier.get();
+
+        this.reaperTask = new ReaperTask(connectionManagers, sleepPeriod);
+
+        e.execute(this.reaperTask);
+
+        exec = e;
     }
 
-    private void markShuttingDown() {
-        shuttingDown = true;
+    private synchronized void cleanupExecutorIfNecessary() {
+        if (exec == null) {
+            return;
+        }
+
+        reaperTask.stop();
+        reaperTask = null;
+        exec.shutdownNow();
+        exec = null;
     }
 
-    @SuppressWarnings("unchecked")
-    @Override
-    public void run() {
-        while (true) {
-            if (shuttingDown) {
-                log.debug("Shutting down reaper thread.");
-                return;
-            }
-            try {
-                Thread.sleep(PERIOD_MILLISECONDS);
-
-                for (Map.Entry<HttpClientConnectionManager, Long> entry : CONNECTION_MANAGERS.entrySet()) {
-                    // When we release connections, the connection manager leaves them
-                    // open so they can be reused.  We want to close out any idle
-                    // connections so that they don't sit around in CLOSE_WAIT.
-                    try {
-                        entry.getKey().closeIdleConnections(entry.getValue(), TimeUnit.MILLISECONDS);
-                    } catch (Exception t) {
-                        log.warn("Unable to close idle connections", t);
+    private static final class ReaperTask implements Runnable {
+        private final Map<HttpClientConnectionManager, Long> connectionManagers;
+        private final long sleepPeriod;
+
+        private volatile boolean stopping = false;
+
+        private ReaperTask(Map<HttpClientConnectionManager, Long> connectionManagers,
+                           long sleepPeriod) {
+            this.connectionManagers = connectionManagers;
+            this.sleepPeriod = sleepPeriod;
+        }
+
+        @Override
+        public void run() {
+            while (true) {
+                if (stopping) {
```

Comment:
Should we just put this in the while? while(!stopping)

Reply:
Done